modin/core/dataframe/algebra/groupby.py

```diff
@@ -383,16 +418,208 @@ def try_filter_dict(agg_func, df):
             Aggregation function.
         df : pandas.DataFrame
             Serialized partition which contains available columns.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
 
         Returns
         -------
-        Callable
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
             Aggregation function that can be safely applied to this particular partition.
         """
         if not isinstance(agg_func, dict):
             return agg_func
-        partition_dict = {k: v for k, v in agg_func.items() if k in df.columns}
-        return lambda grp: grp.agg(partition_dict)
+
+        grp_has_id_level = df.columns.names[0] == cls.ID_LEVEL_NAME
+        # The 'id' level prevents us from a lookup for the original
+        # partition's columns. So dropping the level.
+        partition_columns = frozenset(
+            df.columns.droplevel(0) if grp_has_id_level else df.columns
+        )
+
+        partition_dict = {k: v for k, v in agg_func.items() if k in partition_columns}
+        return cls._build_callable_for_dict(
+            partition_dict, preserve_aggregation_order, grp_has_id_level
+        )
+
+    @classmethod
+    def _build_callable_for_dict(
+        cls, agg_dict, preserve_aggregation_order=True, grp_has_id_level=False
+    ):
+        """
+        Build callable for an aggregation dictionary.
+
+        Parameters
+        ----------
+        agg_dict : dict
+            Aggregation dictionary.
+        preserve_aggregation_order : bool, default: True
+            Whether to manually restore the order of columns for the result specified
+            by the `agg_func` keys (only makes sense when `agg_func` is a dictionary).
+        grp_has_id_level : bool, default: False
+            Whether the frame we're grouping on has intermediate columns
+            (see ``cls.ID_LEVEL_NAME``).
+
+        Returns
+        -------
+        callable(pandas.core.groupby.DataFrameGroupBy) -> pandas.DataFrame
+        """
+        from modin.pandas.utils import walk_aggregation_dict
```

Comment:
Why isn't this at the top of the file?

Reply:
Otherwise, it'll trigger a circular import: [`modin.pandas.__init__` imports `modin.pandas.general`](https://github.com/modin-project/modin/blob/a776db5c77d5b91e5a34f0c0da4a911b8790936a/modin/pandas/__init__.py#L225) that [imports `modin.core.storage_formats. ...`](https://github.com/modin-project/modin/blob/a776db5c77d5b91e5a34f0c0da4a911b8790936a/modin/pandas/general.py#L27) that [imports `PandasQueryCompiler`](https://github.com/modin-project/modin/blob/a776db5c77d5b91e5a34f0c0da4a911b8790936a/modin/core/storage_formats/__init__.py#L17) that [imports `algebra.groupby`](https://github.com/modin-project/modin/blob/a776db5c77d5b91e5a34f0c0da4a911b8790936a/modin/core/storage_formats/pandas/query_compiler.py#L52) that imports `modin.pandas.utils` that again imports `modin.pandas.__init__`.

Added corresponding in-code comment.

Reply:
Should we move this method closer to the place where it is used (`modin/core/storage_formats/pandas/utils.py`, for instance)?

Reply:
Potential usage of this method is not limited by `storage_formats.pandas`. It also could be used by HDK implementations (in order to properly support dict agg) as well as in the `modin.pandas` layer in order to properly preprocess such aggregations.

Reply:
I guess we could move it to something like `modin._internal.utils` (i.e. outside of `modin.pandas`) during some refactoring if we wish... circular imports are bad, but I can live with that.